lib/package.js

```diff
@@ -760,31 +761,50 @@ exports.createMain = function(pkg, pjson, downloadDir) {
     }
 
     if (main) {
-      if (main.startsWith('./'))
-        main = main.substr(2);
-      if (main.endsWith('.js'))
-        main = main.substr(0, main.length - 3);
+      if (main.startsWith('./')) {
+        main = main.substr(2);    
+      }  
+    }
+    else {
+      main = 'index';
     }
 
-    main = main || 'index';
-
-    // try the package.json main
     return new Promise(function(resolve) {
-      mainPath = path.resolve(downloadDir, main + '.js');
+      // Check to see if main points to an actual file or not
+      // i.e. does it include extension.
+      mainPath = path.resolve(downloadDir, main);
+      fs.exists(mainPath, resolve);
+    });
+
+  }).then(function(exists) {
+    if (exists) {
+      var match = /(\.[^.]+)$/.exec(main);
+      if (match) {
+        mainFileExtension = match[1];
+        main = main.slice(0, - mainFileExtension.length);
+      }
+      return exists;     
+    }
+
+    return new Promise(function(resolve) {      
+      mainPath = path.resolve(downloadDir, main + mainFileExtension);
```

Comment:
I can see that you want to reduce it to 2, i.e.
1. if `pjson.main` does not exists, try `pkg.name -> lastNamePart`. If not, try as is
2. If (1) fails and `pjson.main` exists, then try adding extension.
The flow is different, so I don't want to make too big a change (as I see you already refactored the whole thing in `0.17`).

Also, it will normally do no more than 2 check in current implementation anyway, because they will be skipped if the file is found.

If you want, I can do the refactoring, not much of a deal anyway. :)